src/cryptonote_core/service_node_list.cpp

```diff
@@ -954,14 +959,20 @@ namespace service_nodes
     // Cull old history
     //
     {
-      uint64_t checkpoint_immutable_height = m_db->get_checkpoint_immutable_height();
-      uint64_t conservative_height         = block_height < MAX_STATE_HISTORY ? 0 : block_height - MAX_STATE_HISTORY;
+      uint64_t start_height = (block_height < MAX_SHORT_TERM_STATE_HISTORY) ? 0 : block_height - MAX_SHORT_TERM_STATE_HISTORY;
+      auto it =
+          std::lower_bound(m_state_history.begin(),
+                           m_state_history.end(),
+                           start_height,
+                           [](state_t const &state, uint64_t start_height) { return state.height < start_height; });
 
-      uint64_t const cull_height = (checkpoint_immutable_height == 0)
-                                       ? conservative_height
-                                       : std::min(checkpoint_immutable_height, conservative_height);
-      while (!m_state_history.empty() && m_state_history.front().height < cull_height)
-        m_state_history.pop_front();
+      for (; it != m_state_history.end() && it->height <= start_height; it++)
+      {
+        bool long_term_storage = (it->height % 10000 == 0);
+        if (long_term_storage) continue;
+        it = m_state_history.erase(it);
+        it--;
```

Comment:
This `--` is not valid in some cases: if `erase` returns the new head of the deque then the `--` is going to put `it` in an invalid state (i.e. before the beginning).

Better approach:
```C++
while (it != m_state_history.end() && it->height <= start_height) {
    if (it->height % 10000 == 0)
        it++;
    else
        it = m_state_history.erase(it);
}
```

Reply:
Can it return an iterator to begin? According to docs (https://en.cppreference.com/w/cpp/container/vector/erase),  if you erase the last element you get the end iterator and end() - 1 gives you begin()?

Also the advantage to having a deque here isn't clear anymore, so was also gonna change it to a vector. Was running this in the office to see what the cost of this was on mainnet but was still syncing before I left so gonna wait for the results of that